tephra-core/src/main/java/org/apache/tephra/util/TxUtils.java

```diff
@@ -163,8 +167,22 @@ public static long getPruneUpperBound(Transaction tx) {
 
     long maxInvalidTx =
       tx.getInvalids().length > 0 ? tx.getInvalids()[tx.getInvalids().length - 1] : Transaction.NO_TX_IN_PROGRESS;
+
+    // An invalid transaction can be used up to its max lifetime for data writes, hence we cannot prune an invalid
+    // transaction until it exhausts its max lifetime.
+    long elapsedTime = currentTimeMillis - TxUtils.getTimestamp(maxInvalidTx);
+    long invalidTxBound;
+    if (elapsedTime > txMaxLifetimeMillis) {
+      // maxInvalidTx is past its max lifetime
+      invalidTxBound = maxInvalidTx;
+    } else {
+      // Reduce the lifetime of maxInvalidTx, so that it cannot be used for any more writes
+      long remainingTime = txMaxLifetimeMillis - elapsedTime;
+      invalidTxBound = maxInvalidTx - remainingTime * TxConstants.MAX_TX_PER_MS - 1;
```

Comment:
Good point. Now that I think about it, we can further simplify the method `TxUtils.getPruneUpperBound()` to simply return `(currentTimeMillis - txMaxLifetimeMillis) * TxConstants.MAX_TX_PER_MS - 1`, irrespective of whether there are in-progress transactions or not. Since no transaction (in-progress or invalid) can be used for writes after `txMaxLifetimeMillis`.

What do you say?